File: zZ/java/compare-version-numbers.java
/**
 * Compare Version Numbers
 * 
 * Compare two version numbers version1 and version2.
 * If version1 > version2 return 1; if version1 <version2 return -1;otherwise return 0.
 * 
 * You may assume that the version strings are non-empty and contain only digits and the . character.
 * The . character does not represent a decimal point and is used to separate number sequences.
 * For instance, 2.5 is not "two and a half" or "half way to version three", it is the fifth second-level revision of the second first-level revision.
 * 
 * Example 1:
 * 
 * Input: version1 = "0.1", version2 = "1.1"
 * Output: -1
 * Example 2:
 * 
 * Input: version1 = "1.0.1", version2 = "1"
 * Output: 1
 * Example 3:
 * 
 * Input: version1 = "7.5.2.4", version2 = "7.5.3"
 * Output: -1
 * Credits:
 * Special thanks to @ts for adding this problem and creating all test cases.
 */

public class Solution {
    public int compareVersion(String version1, String version2) {
        for (int index1 = 0, index2 = 0; index1 < version1.length() || index2 < version2.length(); ++index1, ++index2) {
            int subVersion1 = 0;
            while (index1 < version1.length() && version1.charAt(index1) != '.') {
                subVersion1 = 10 * subVersion1 + version1.charAt(index1++) - '0';
            }

            int subVersion2 = 0;
            while (index2 < version2.length() && version2.charAt(index2) != '.') {
                subVersion2 = 10 * subVersion2 + version2.charAt(index2++) - '0'; 
            }

            if (subVersion1 != subVersion2) {
                return subVersion1 > subVersion2 ? 1 : -1;
            }
        }

        return 0;
    }
}

File: data-structure-and-algorithm/cpp/compare-version-numbers.cpp
// 165. Compare Version Numbers
// Difficulty: Easy

// Compare two version numbers version1 and version2.
// If version1 > version2 return 1, if version1 < version2 return -1, otherwise return 0.

// You may assume that the version strings are non-empty and contain only digits and the . character.
// The . character does not represent a decimal point and is used to separate number sequences.
// For instance, 2.5 is not "two and a half" or "half way to version three",
// it is the fifth second-level revision of the second first-level revision.

// Here is an example of version numbers ordering:

// 0.1 < 1.1 < 1.2 < 13.37

// Time:  O(n)
// Space: O(1)

class Solution {
public:
  int compareVersion(string version1, string version2) {
    const int n1 = version1.length(), n2 = version2.length();
    for (int i = 0, j = 0; i < n1 || j < n2; ++i, ++j) {
      int v1 = 0, v2 = 0;
      while (i < n1 && version1[i] != '.') {
        v1 = v1 * 10 + version1[i++] - '0';
      }
      while (j < n2 && version2[j] != '.') {
        v2 = v2 * 10 + version2[j++] - '0';
      }
      if (v1 != v2) {
        return v1 > v2 ? 1 : -1;
      }
    }
    return 0;
  }
};
